File: src/FSharpVSPowerTools/HighlightUsageFilterProvider.cs
using System.Diagnostics;
using System.ComponentModel.Composition;
using Microsoft.VisualStudio;
using Microsoft.VisualStudio.Editor;
using Microsoft.VisualStudio.OLE.Interop;
using Microsoft.VisualStudio.Text.Editor;
using Microsoft.VisualStudio.Text.Operations;
using Microsoft.VisualStudio.TextManager.Interop;
using Microsoft.VisualStudio.Utilities;
using Microsoft.VisualStudio.Shell;
using FSharpVSPowerTools.ProjectSystem;
using FSharpVSPowerTools.HighlightUsage;
using Microsoft.VisualStudio.Text;
using Microsoft.VisualStudio.Text.Tagging;

namespace FSharpVSPowerTools
{
    [Export(typeof(IVsTextViewCreationListener))]
    [ContentType("F#")]
    [TextViewRole(PredefinedTextViewRoles.Editable)]
    internal class HighlightUsageFilterProvider : IVsTextViewCreationListener
    {
        [Import]
        internal IVsEditorAdaptersFactoryService editorFactory = null;

        [Import(typeof(SVsServiceProvider))]
        internal System.IServiceProvider serviceProvider = null;

        [Import]
        internal IViewTagAggregatorFactoryService tagAggregator = null;

        public void VsTextViewCreated(IVsTextView textViewAdapter)
        {
            var textView = editorFactory.GetWpfTextView(textViewAdapter);
            if (textView == null) return;

            var generalOptions = Setting.getGeneralOptions(serviceProvider);
            if (generalOptions == null || !generalOptions.HighlightUsageEnabled) return;
            
            AddCommandFilter(textViewAdapter,
                new HighlightUsageFilter(textView, tagAggregator.CreateTagAggregator<TextMarkerTag>(textView)));            
        }

        private static void AddCommandFilter(IVsTextView viewAdapter, HighlightUsageFilter commandFilter)
        {
            if (!commandFilter.IsAdded)
            {
                // Get the view adapter from the editor factory
                IOleCommandTarget next;
                int hr = viewAdapter.AddCommandFilter(commandFilter, out next);

                if (hr == VSConstants.S_OK)
                {
                    commandFilter.IsAdded = true;
                    // You'll need the next target for Exec and QueryStatus
                    if (next != null) commandFilter.NextTarget = next;
                }
            }
        }
    }
}

File: src/FSharpVSPowerTools.Pure/Commands/HighlightUsageFilterProvider.fs
module FSharpVSPowerTools.HighlightUsageFilterProvider

open System.Diagnostics
open System.ComponentModel.Composition
open Microsoft.VisualStudio
open Microsoft.VisualStudio.Editor
open Microsoft.VisualStudio.OLE.Interop
open Microsoft.VisualStudio.Text.Editor
open Microsoft.VisualStudio.Text.Operations
open Microsoft.VisualStudio.TextManager.Interop
open Microsoft.VisualStudio.Utilities
open Microsoft.VisualStudio.Shell
open FSharpVSPowerTools.ProjectSystem
open FSharpVSPowerTools.HighlightUsage
open Microsoft.VisualStudio.Text
open Microsoft.VisualStudio.Text.Tagging
open System.ComponentModel.Composition
open System


[<Export(typeof<IVsTextViewCreationListener>)>]
[<ContentType "F#">]
[<TextViewRole(PredefinedTextViewRoles.Editable)>]
type HighlightUsageFilterProvider [<ImportingConstructor>]
    ( [<Import (typeof<SVsServiceProvider>)>] 
        serviceProvider :   IServiceProvider                    ,
        editorFactory   :   IVsEditorAdaptersFactoryService     ,
        tagAggregator   :   IViewTagAggregatorFactoryService    ) =
    

    static member AddCommandFilter (viewAdapter:IVsTextView, commandFilter:HighlightUsageFilter) =
        if not commandFilter.IsAdded then
            let mutable next = Unchecked.defaultof<IOleCommandTarget>
            let hr = viewAdapter.AddCommandFilter(commandFilter, &next)

            if hr = VSConstants.S_OK then
                commandFilter.IsAdded <- true
                // You'll need the next target for Exec and QueryStatus
                if isNotNull next then commandFilter.NextTarget <- next


    interface IVsTextViewCreationListener with

        member __.VsTextViewCreated textViewAdapter =
            unitMaybe {
                let! textView = editorFactory.TryGetWpfTextView textViewAdapter
                let! generalOptions = Setting.tryGetGeneralOptions serviceProvider
                if not generalOptions.HighlightUsageEnabled then return! None else

                HighlightUsageFilterProvider.AddCommandFilter ( textViewAdapter, 
                        new HighlightUsageFilter (textView, tagAggregator.CreateTagAggregator<TextMarkerTag> textView))
            } 

